tests/test_eepromfs.c

```diff
@@ -0,0 +1,93 @@
+void test_eepromfs(TestContext *ctx) {
+    const int blocks = eeprom_total_blocks();
+    if (blocks == 0) {
+        SKIP("EEPROM not found; skipping eepfs tests");
+    }
+    if (IN_EMULATOR) {
+        // Gracefully handle broken default EEPROM in cen64
+        uint64_t eeprom_block = 0;
+        uint8_t * eeprom_bytes = (uint8_t *)&eeprom_block;
+        eeprom_write(0, eeprom_bytes);
+        eeprom_read(0, eeprom_bytes);
+        if (eeprom_block != 0) {
+            SKIP("cen64 defaults to broken 4K EEPROM implementation");
```

Comment:
https://github.com/n64dev/cen64/blob/a56fa4ba41da2c15cb054c063082988fb66780c0/si/controller.c#L124-L132

Reply:
A better option that could be investigated would be to make a PR for CEN64 that fixes this issue there.

Reply:
I completely agree that cen64's behavior is incorrect, but the very existence of the hack means that someone found it to be necessary under some circumstance (the details of which are unfortunately missing).

I implemented this workaround in the spirit of [Chesterton's Fence](https://www.chesterton.org/taking-a-fence-down/):
>In the matter of reforming things, as distinct from deforming them, there is one plain and simple principle; a principle which will probably be called a paradox. There exists in such a case a certain institution or law; let us say, for the sake of simplicity, a fence or gate erected across a road. The more modern type of reformer goes gaily up to it and says, “I don’t see the use of this; let us clear it away.” To which the more intelligent type of reformer will do well to answer: “If you don’t see the use of it, I certainly won’t let you clear it away. Go away and think. Then, when you can come back and tell me that you do see the use of it, I may allow you to destroy it.

Since I do not know precisely why this hack was implemented in cen64, I was afraid to touch it in case my attempt to "fix it" actually ended up breaking something else in unexpected ways.

Reply:
I do not maintain cen64, but in general for emulators that focus on accuracy its better to fix bugs even if that means that temporarily games get broken -- it would prompt people to find and do the correct fix.

Reply:
It looks like mupen64plus has the same hack; it defaults to a 4K EEPROM in the Joybus command response, even if the cart doesn't actually have an EEPROM.

https://github.com/mupen64plus/mupen64plus-core/blob/db56853707ca6f4223adf5e7f0b7725f90d3a0a3/src/main/main.c#L1703-L1724

https://github.com/mupen64plus/mupen64plus-core/blob/db56853707ca6f4223adf5e7f0b7725f90d3a0a3/src/device/cart/cart.c#L45-L52

I tried to find documentation on how this is "supposed" to work, but I can't find a definitive answer.

https://n64brew.dev/wiki/Joybus_Protocol#0x00_-_Info does not indicate that there is an identifier for "No EEPROM present".

Libdragon assumes that no EEPROM is present if the identifier is not 4K EEPROM (`0x0080`) and is not 16K EEPROM (`0x00C0`), which seems reasonable.

I may need to build a test ROM that logs out the Joybus response on real hardware to get to the bottom of this.

Reply:
A real console responds with:

Save Type | Rx Bytes
--- | ---
EEPROM 4K | `0x008000`
EEPROM 16K | `0x00C000`
All others | `0xFFFFFF`

I'll write up some PRs for Mupen64Plus and cen64 to respond accurately to EEPROM status inquiries.

Thanks for your feedback @rasky  and @sp1187

Reply:
https://github.com/n64dev/cen64/pull/203

Reply:
https://github.com/mupen64plus/mupen64plus-core/pull/878

Reply:
Cen64's behavior when no EEPROM is specified is straight-up broken. Mupen64Plus just always has a 4K EEPROM unless 16K EEPROM is specified, which is not accurate but is not exactly harmful either.